### PR TITLE
fix #269672: Show selected notes on the piano control

### PR DESCRIPTION
**Pull request by mattmcclinch**
Alternative to #3488.

### CODE REVIEW

File: mscore/pianotools.cpp [hunk 1]
```diff
@@ -167,6 +168,25 @@ void HPiano::releasePitch(int pitch)
...
+//   changeSelection
+//---------------------------------------------------------
+
+void HPiano::changeSelection(Selection selection)
+      {
+      for (PianoKeyItem* key : keys) {
+            key->setHighlighted(false);
+            key->setSelected(false);
+            }
+      for (Note*n : selection.uniqueNotes()) {
```
Comment (lasconic) [inline mscore/pianotools.cpp]: a space is missing here.

File: mscore/pianotools.cpp [hunk 2]
```diff
@@ -167,6 +168,25 @@ void HPiano::releasePitch(int pitch)
...
+            keys[n->pitch()-_firstKey]->setSelected(true);
```
Comment (lasconic) [inline mscore/pianotools.cpp]: Please use spaces around -

File: mscore/pianotools.cpp [hunk 3]
```diff
@@ -167,6 +168,25 @@ void HPiano::releasePitch(int pitch)
...
+            for (Note* other : n->chord()->notes())
+                  keys[other->pitch()-_firstKey]->setHighlighted(true);
```
Comment (lasconic) [inline mscore/pianotools.cpp]: same here, spaces around -

**Comment by mattmcclinch** [comment]
Okay, I added the missing spaces.

When I was testing this patch in 3.0, I noticed that the selected piano key did not change if I added an accidental to the selected note. I investigated, and found this in libmscore/undo.cpp:
`
//---------------------------------------------------------
//   SaveState
//---------------------------------------------------------

SaveState::SaveState(Score* s)
   : undoInputState(s), redoInputState(s->inputState())
      {
      score = s;
//      redoSelection  = score->selection();
      }

void SaveState::undo(EditData*)
      {
      redoInputState = score->inputState();
//      redoSelection  = score->selection();
      score->setInputState(undoInputState);
//      score->setSelection(undoSelection);
      }

void SaveState::redo(EditData*)
      {
      undoInputState = score->inputState();
//      undoSelection  = score->selection();
      score->setInputState(redoInputState);
//      score->setSelection(redoSelection);
      }
`

Is there any reason these lines are commented out?

**Comment by lasconic** [comment]
No idea, @wschweer ?

**Comment by lasconic** [comment]
I will just merge it  as it in 3.0 if possible and let's solve the problem later with @wschweer.

**Comment by wschweer** [comment]
Its difficult to store the selection state in the undo/redo history as its hard to avoid dangling pointers to elements which are deleted during layout. So until there is a better solution i commented out the functionality.

**Comment by mattmcclinch** [comment]
Interesting. Does version 2.x have this dangling pointer problem?

**Comment by mattmcclinch** [comment]
Did you ever have a look at #3476? It addresses a problem similar to what you describe. It may be a step toward the solution you are looking for.

**Pull request merged**